IPIP/0332-gateway-error-handling.md

```diff
@@ -0,0 +1,72 @@
+# IPIP 0000: Streaming Error Handling in HTTP Gateways
+
+- Start Date: 2022-10-12
+- Related Issues:
+  - [ipfs/kubo/pull/9333](https://github.com/ipfs/kubo/pull/9333)
+  - [mdn/browser-compat-data/issues/14703](https://github.com/mdn/browser-compat-data/issues/14703)
+
+## Summary
+
+Ensure streaming error handling in web gateways is clear and consistent.
+
+## Motivation
+
+Web gateways provide different functionalities where users can download files.
+The download of this files is streamed from the server to the client using HTTP.
+However, there is no good way of presenting to the client an error that happens
+during the stream.
+
+For example, if during the download of a TAR file, the server detects some error
+and is not able to continue, the user can get a valid, yet incomplete TAR. However,
+the user will not know that the TAR is incomplete. By introducing consistent error
+handling, the server attempts to notify the user.
+
+## Detailed design
+
+If the server encounters an error before streaming the contents to the client,
+the server must fail with the respective `4xx`  or `5xx` HTTP status code (no change).
+
+If the server encounters an error while streaming the contents, the server must
+force-close the HTTP stream to the user. This way, the user will receive a
+network error, making it clear that the downloaded file is not valid.
+In addition, the server must set an header `X-On-Error: reset` to indicate that
+connection will be reset if an error occurs.
+
+The client can opt-out of this behavior by sending a `X-Error-Behavior: trailer`.
+In that case, the server will not force-close the HTTP stream. Instead, the server
+will send the trailing header `X-Stream-Error` header with the error message.
```

Comment:
As usual, reminding that Trailers can't be read by JS in web browsers: https://github.com/mdn/browser-compat-data/issues/14703